### ISSUE TITLE
"x" =~ /\A(?>(?:(?:)A|B|C?x))\z/; no longer matches in blead

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#12006]
Migrated from [rt.perl.org#111842](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=111842) (status was 'resolved')

Searchable as RT111842$

**Comment by p5pRT** [comment]
### From @nwc10

#### Created by @nwc10



"x" =~ /\\A\(?>\(?&#8203;:\(?&#8203;:\)A|B|C?x\)\)\\z/; no longer matches in blead\.

Zefram reduced the test case down from this BBC report&#8203;:

http&#8203;://www\.cpantesters\.org/cpan/report/4143b066\-70b1\-11e1\-9c4c\-03c8ecbfb8b4

I've bisected it with&#8203;:

\.\./perl/Porting/bisect\.pl \-\-target=miniperl \-\-start  v5\.14\.0 \-e 'die unless "x" =~ /\\A\(?>\(?&#8203;:\(?&#8203;:\)A|B|C?x\)\)\\z/;'

Culprit reported as&#8203;:

HEAD is now at 2f137bb make test\.pl show test number and name in failure diagnostics output
good \- zero exit from \./miniperl \-Ilib \-e die unless "x" =~ /\\A\(?>\(?&#8203;:\(?&#8203;:\)A|B|C?x\)\)\\z/;
fab2782b37b5570d7f8f8065fd7d18621117ed49 is the first bad commit
commit fab2782b37b5570d7f8f8065fd7d18621117ed49
Author&#8203;: Yves Orton \<demerphq@&#8203;gmail\.com>
Date&#8203;:   Sun Feb 19 21&#8203;:32&#8203;:05 2012 \+0100

&nbsp;    rework how the trie logic handles the newer EXACT nodetypes

&nbsp;    This cleans up and simplifies and extends how the trie
&nbsp;    logic interacts with the new node types\. This change ultimately
&nbsp;    makes the EXACTFU\, EXACTFU\_SS\, EXACTFU\_NO\_TRIE \(renamed to
&nbsp;    EXACTFU\_TRICKYFOLD\) work properly with the trie engine regardless
&nbsp;    of whether the string is utf8 or latin1\.

&nbsp;    This patch depends on the following&#8203;:

&nbsp;        EXACT              => utf8 or "binary" text

&nbsp;        EXACTFU            => either pre\-folded utf8\, or latin1 that has to be folded as though it was utf8
&nbsp;        EXACTFU\_SS         => special case of EXACTFU to handle \\xDF/ss \(affects latin1 treatment\)
&nbsp;        EXACTFU\_TRICKYFOLD => special case of EXACTFU to handle tricky non\-latin1 fold rules

&nbsp;        EXACTF             => "old style fold logic" untriable nodetype
&nbsp;        EXACTFA            => \(currently\) untriable nodetype
&nbsp;        EXACTFL            => \(currently\) untriable nodetype

&nbsp;    See the comments in regcomp\.sym for these fold types\.

&nbsp;    This patch involves a number of distinct\, but related parts\. Starting
&nbsp;    from compilation&#8203;:

&nbsp;    \* Simplify how we detect a triable sequence given the new nodetypes\,
&nbsp;      this also probably fixed some "bugs" in how we detected certain
&nbsp;      sequences\, like /||foo|bar/\.

&nbsp;    \* Simplify how we read EXACTFU nodes under utf8 by removing the now
&nbsp;      redundant folding logic \(EXACTFU nodes under utf8 are prefolded\)\.
&nbsp;      Also extend this logic to handle latin1 patterns properly \(in
&nbsp;      conjunction with  other changes\)

&nbsp;    \* Part of the problems associated with EXACTFU\_SS and EXACTFU\_TRICKYFOLD
&nbsp;      have to do with how the trie logic interacts with the minlen logic\.
&nbsp;      This change handles both by pessimising the minlen when encounting
&nbsp;      these nodetypes\. One observation is that the minlen logic is basically
&nbsp;      broken\, and works only because it conflates bytes and codepoints in
&nbsp;      such a way that we more or less always get a value small enough that things work out
&nbsp;      anyway\. Fixing that is properly is the job of another patch\.

&nbsp;    \* Part of the problem of doing folding under unicode rules is that
&nbsp;      there are a lot of foldings possible\, some with strange rules\. This
&nbsp;      means that the bitmap logic does not work correctly in all cases\,
&nbsp;      as we currently do not have any way to populate it properly\.
&nbsp;      So this patch disables the bitmap entirely when folding is involved
&nbsp;      until that is fixed\.

&nbsp;    The end result of this is&#8203;: we can TRIE/AHOCORASICK any sequence of
&nbsp;    EXACT\, or EXACTFU \(ish\) nodes\, regardless of utf8 or not\, but we disable
&nbsp;    the bitmap when folding\.

&nbsp;    A note for follow up relating to this patch is that the way EXACTFU\_XXX
&nbsp;    nodes are currently dealt with we wont build the "maximal" trie because
&nbsp;    of their presence\, instead creating a "jumptrie" consisting of either a
&nbsp;    leading EXACTFU node followed by a EXACTFU\_XXX node\, or vice versa\. We
&nbsp;    should eventually address that\.

:100644 100644 5c380fffd5b22ad4543da79762d563540322766e c549dc9784646f61155696dc61f0ee008d3b4f90 M      embed\.fnc
:100644 100644 9fdf91b7ce8e13e7c716357d84b6071eb8ba89df 1d1e598733a375240fc7b8380ada328731ba04a7 M      embed\.h
:100644 100644 dd3fd587de6aedae789d87926c3cc84218f78e3f b811e6b22fc4179209ce91a178b73e33cf55250a M      proto\.h
:100644 100644 7a3433a072c5a5f3ca435b808ddaa2bedd94bb05 e3da6e935105850c14e15c954a88d0ef12fc5ae1 M      regcomp\.c
:100644 100644 f0ee7019915b03026b7e5b4194f290cfe91758b9 e6be6ac2924a50398432e28388503bc6ccd71529 M      regcomp\.sym
:100644 100644 76eb7f9bedc63996c887d080a47ec91ade216346 8ccb6f7032e389593339589938d74d1e324c3cea M      regexec\.c
:100644 100644 994159342fba98343ea1edf54b66c6bd7b48512f 55612438824762d0b3879115903957e7ff36f9b4 M      regnodes\.h
:040000 040000 e3e998b3f5b780941d607fa28e725a7d633b10dc 6b3a1e00bc5da26ff92d829c92bac3712691b48e M      t
bisect run success
That took 495 seconds


Sadly I think this needs to be a 5\.16\.0 blocker\, at least until we
understand why it's failing\.

Nicholas Clark

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=high

Site configuration information for perl 5.15.8:

Configured by nick at Mon Mar 19 12:12:40 CET 2012.

Summary of my perl5 (revision 5 version 15 subversion 8) configuration:
  Derived from: b6780f634d3f541f3dfa78a8674e8b081bd27c4f
  Platform:
    osname=linux, osvers=2.6.32-5-amd64, archname=x86_64-linux
    uname='linux gcc20 2.6.32-5-amd64 #1 smp wed may 18 23:13:22 utc 2011 x86_64 gnulinux '
    config_args='-Dusedevel=y -Dcc=ccache gcc -Dld=gcc -Ubincompat5005 -Uinstallusrbinperl -Dcf_email=nick@ccl4.org -Dperladmin=nick@ccl4.org -Dinc_version_list=  -Dinc_version_list_init=0 -DDEBUGGING -Doptimize=-g -Uusethreads -Uuselongdouble -Uuse64bitall -Uusemymalloc -Duseperlio -Dprefix=~/Sandpit/snap-v5.15.8-179-gb6780f6 -Uusevendorprefix -Uvendorprefix=~/Sandpit/snap-v5.15.8-179-gb6780f6 -Uuserelocatableinc -Umad -Accflags=-DNO_MATHOMS -de'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='ccache gcc', ccflags ='-DNO_MATHOMS -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-g',
    cppflags='-DNO_MATHOMS -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.4.5', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib/../lib /usr/lib/../lib /lib /usr/lib /lib64 /usr/lib64
    libs=-lnsl -ldl -lm -lcrypt -lutil -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.11.2.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.11.2'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -g -L/usr/local/lib -fstack-protector'

Locally applied patches:
    


@INC for perl 5.15.8:
    lib
    /home/nick/Sandpit/snap-v5.15.8-179-gb6780f6/lib/perl5/site_perl/5.15.8/x86_64-linux
    /home/nick/Sandpit/snap-v5.15.8-179-gb6780f6/lib/perl5/site_perl/5.15.8
    /home/nick/Sandpit/snap-v5.15.8-179-gb6780f6/lib/perl5/5.15.8/x86_64-linux
    /home/nick/Sandpit/snap-v5.15.8-179-gb6780f6/lib/perl5/5.15.8
    .


Environment for perl 5.15.8:
    HOME=/home/nick
    LANG=en_US.UTF-8
    LANGUAGE=en_US:en
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/nick/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @rjbs

If this cannot be easily and safely fixed\, it should be reverted until 5\.17\, before 5\.15\.9\.

But Yves\, the author of the original commit\, says&#8203;:

&nbsp;  I know the cause
&nbsp;  i will let you know more later
&nbsp;  it probably can be easily and safely fixed\. 

We look forward to hearing more from him today\.

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

If this cannot be easily and safely fixed\, it should be reverted until 5\.17\, before 5\.15\.9\.

But Yves\, the author of the original commit\, says&#8203;:

&nbsp;  I know the cause
&nbsp;  i will let you know more later
&nbsp;  it probably can be easily and safely fixed\. 

We look forward to hearing more from him today\.

**Comment by p5pRT** [comment]
@rjbs - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @demerphq

On 19 March 2012 15&#8203;:26\, Ricardo SIGNES via RT \<perlbug\-comment@&#8203;perl\.org> wrote&#8203;:
> If this cannot be easily and safely fixed\, it should be reverted until 5\.17\, before 5\.15\.9\.
>
> But Yves\, the author of the original commit\, says&#8203;:
>
>  I know the cause
>  i will let you know more later
>  it probably can be easily and safely fixed\.
>
> We look forward to hearing more from him today\.

Hi\, I am looking into a fix now\. If I cant do a fix before I leave the
office \(its after COB\) then I will mail you to let you know to revert\.

Yves

\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @demerphq

On 19 March 2012 18&#8203;:06\, demerphq \<demerphq@&#8203;gmail\.com> wrote&#8203;:
> On 19 March 2012 15&#8203;:26\, Ricardo SIGNES via RT \<perlbug\-comment@&#8203;perl\.org> wrote&#8203;:
>> If this cannot be easily and safely fixed\, it should be reverted until 5\.17\, before 5\.15\.9\.
>>
>> But Yves\, the author of the original commit\, says&#8203;:
>>
>>  I know the cause
>>  i will let you know more later
>>  it probably can be easily and safely fixed\.
>>
>> We look forward to hearing more from him today\.
>
> Hi\, I am looking into a fix now\. If I cant do a fix before I leave the
> office \(its after COB\) then I will mail you to let you know to revert\.

I wanted to give an update since it is taking longer than expected\.

I can replicate this bug\, and related bugs in 5\.12\.4 with minor
modifications of the pattern&#8203;:

$ perl \-E'say "D" =~ /\\A\(?>\(?&#8203;:A|\(?&#8203;:\)B|\(?&#8203;:\)D|E|C?x\)\)\\z/ ? "ok&#8203;: $&" : "not ok"'
not ok
$ perl \-E'say "D" =~ /\\A\(?>\(?&#8203;:A|B|D|E|C?x\)\)\\z/ ? "ok&#8203;: $&" : "not ok"'
ok&#8203;: D

So clearly this bug has been around a while\. IMO this bug should be
seen as an outright blocker and not as a "revert if it cant be fixed"
bug\. Id like to say more but I am still not sure exactly what is the
cause here\. Although I do think I am close\. \(Dave M if you are reading
and have a eureka then please let me know\, this involves some of the
trie code that you reworked way back\.\)

It is late here\, and I cannot guarantee that I can find a proper fix
tonite\, although I plan to try\. If I cannot solve it tonite I reckon I
can get it solved in the next few days\.

Note\, to be clear\, a simple revert will not fix this problem\, the
patch involved only made the problem easier to see\, but it has existed
for a while\.

cheers\,
Yves


\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @abigail

This was resolved in commit a40630bf5

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

This was resolved in commit a40630bf5

**Comment by p5pRT** [comment]
@abigail - Status changed from 'open' to 'resolved'
